src/main.cc

```diff
@@ -465,47 +465,60 @@ int main( int argc, char ** argv )
-    locale = QLocale( cfg.preferences.interfaceLanguage );
-  }
-
-  QLocale::setDefault( locale );
-  QApplication::setLayoutDirection( locale.textDirection() );
-
-  // Load translations, note some quirks:
-  // * For Windows, windeployqt will combine multiple qt modules translations into `qt_*` thus no `qtwebengine_*` exists
-  // * Only try loading qt & webengine translator GD's translations success to avoid inconsistency
-  // * Use the QLocale based loading QTranslator::load
-
-  QTranslator gd_ts;
-
-  QTranslator qt_ts;
-  QTranslator webengine_ts;
+  // Update default locale
+  if ( !cfg.preferences.interfaceLanguage.isEmpty() && QLocale().name() != cfg.preferences.interfaceLanguage ) {
+    QLocale::setDefault( QLocale( cfg.preferences.interfaceLanguage ) );
+  }
+  QApplication::setLayoutDirection( QLocale().textDirection() );
+
+  { /// Translations
+    auto loadTranslation_qlocale = []( QTranslator & qtranslator,
+                                       const QString & filename,
+                                       const QString & prefix,
+                                       const QString & directory ) -> bool {
+      if ( qtranslator.load( QLocale(), filename, prefix, directory ) ) {
+        qDebug() << "TS loaded: " << qtranslator.filePath();
+        return true;
+      }
+      else {
+        qDebug() << "TS failed to load: " << QLocale().uiLanguages() << filename << prefix << " from " << directory;
+        return false;
+      }
+    };
+
+    auto * gd_ts        = new QTranslator( &app );
+    auto * qt_ts        = new QTranslator( &app );
+    auto * webengine_ts = new QTranslator( &app );
+
+    // For GD's translations,
+    // If interfaceLanguage is explictly set, uses filename based loading, because QLocale sometimes doesn't match GD's translation file name
+    // Only load qt & webengine translators if GD's translation loading succeed to avoid inconsistency
+    // TODO: The QLocale based method sometimes does not work https://github.com/xiaoyifang/goldendict-ng/issues/2120
+    // GD's locale names may mismatch system locale and the return of default QLocale().name() is slightly different across platforms.
+    if ( cfg.preferences.interfaceLanguage.isEmpty() ?
+           loadTranslation_qlocale( *gd_ts, QString(), QString(), Config::getLocDir() ) :
+           gd_ts->load( cfg.preferences.interfaceLanguage, Config::getLocDir() ) ) {
+      QCoreApplication::installTranslator( gd_ts );
+      qDebug() << "TS loaded: " << gd_ts->filePath();
+
+      // For macOS bundle, the QLibraryInfo::TranslationsPath is overriden by GD.app/Contents/Resources/qt.conf
+
+      // For Windows, windeployqt will combine multiple qt modules translations into `qt_*` thus no `qtwebengine_*` exists
+      // qtwebengine loading will fail on Windows.
+
+      if ( loadTranslation_qlocale( *qt_ts, "qt", "_", QLibraryInfo::path( QLibraryInfo::TranslationsPath ) ) ) {
+        QCoreApplication::installTranslator( qt_ts );
+      }
```

Comment:
Will qt load this by default itself?

Reply:
No.

Reply:
what's the purpose of this qt_* ?

Reply:
Various random strings, including error strings
- https://github.com/qt/qttranslations/blob/dev/translations/qtbase_zh_CN.ts
- https://github.com/qt/qttranslations/blob/dev/translations/qtwebengine_zh_CN.ts